// kali/lib/workerpool.js
var _ = require("underscore");
var cluster = require('cluster');

// Coffee-script's bind
var __bind = function (fn, self) {
  return function () {
    return fn.apply(self, arguments);
  };
}

var WorkerPool = (function () {

  // Constructor
  function WorkerPool (options) {
    
    this.binds = ["start", "kill", "bury", "spawn", "reap", "reaper"];
    for(var i in this.binds){
      __bind(this[this.binds[i]], this);
    }
    
    this.workers = {};
    
    cluster.on('death', this.bury);
    cluster.on('death', this.spawn);
  };
  
  
  WorkerPool.prototype.list = function () {

    return _.keys(this.workers);
  };
  
  WorkerPool.prototype.bury = function (worker) {

    delete this.workers[worker.pid];
    if (_.keys(this.workers).length === 0) {
      return process.exit(0);
    }
  };
  
  
  WorkerPool.prototype.reaper = function () {

    var self = this;
    return this.reapTimer = setInterval(function () {

      return self.reap();
    }, this.options.reapTime)
  }
  
  
})();

module.exports = WorkerPool;
